### PR TITLE
Language own letter for file kind

### LINKED ISSUE #317: Conflicting in kind letter F
Bad news.

Incompatible change may be needed to fix this issue.

F used as file tag in ctags main part, is also used for other kinds in some parsers.
Here is an example of ruby.

> % cat /tmp/foo.rb 
cat /tmp/foo.rb 
class ClassExample
        def self.class_method
            puts "in class_method"
        end
end        
% ./ctags --extra=+f  /tmp/foo.rb ; cat tags | grep -v '!' | grep F
./ctags --extra=+f  /tmp/foo.rb ; cat tags | grep -v '!' | grep F
class_method	/tmp/foo.rb	/^        def self.class_method$/;"	F	class:ClassExample
foo.rb	/tmp/foo.rb	1;"	F

ctags --list-kinds says:

ObjectiveC
    F  Object field
Ruby
    F  singleton methods
Rust
    F  A method
SQL
    F  record fields

### PR DESCRIPTION
**Pull request by masatake**
This fixes #317.

Cod in this PR does:
* permit defining language own letter for file kind,
* let some parsers like Ruby using F for their own kinds use ! as letter for file kind,
* introduce long long flag fileKind for --langdef option; xcmd and regex parsers can use it,
* introduce --list-file-lind option,
* check the letter confliction about file kind when initializing parsers,
* introduce more tmain tests, and
* describe the changes in command line of ctags.

### CODE REVIEW

File: main/parse.h [hunk 1]
```diff
@@ -75,6 +93,7 @@ typedef struct {
 	char* name;                    /* name of language */
 	kindOption* kinds;             /* tag kinds handled by parser */
 	unsigned int kindCount;        /* size of `kinds' list */
+	char fileKind;		        /* override letter for file kind */
```
Comment (vhda) [inline main/parse.h]: Could you please align the indent of this comment? OCD alert! ;)

File: main/parse.c [hunk 2]
```diff
@@ -97,6 +99,19 @@ extern const char *getLanguageName (const langType language)
 	return result;
 }
 
+extern const char getLanguageFileKind (const langType language)
+{
+	char kind;
+
+	Assert (0 <= language  &&  language < (int) LanguageCount);
```
Comment (vhda) [inline main/parse.c]: `language >= 0` would be more readable, don't you agree?

File: main/parse.c [hunk 3]
```diff
@@ -1158,6 +1189,28 @@ static void lazyInitialize (langType language)
 /*
 *   Option parsing
 */
+#define COUNT(D) (sizeof(D)/sizeof(D[0]))
+
+static void lang_def_flag_file_kind_long (const char* const optflag, const char* const param, void* data)
```
Comment (vhda) [inline main/parse.c]: Why no camel back?

**Comment by vhda** [comment]
I like this approach because it allows backward compatibility in external tools using ctags, which would not happen if we modified the kind letters currently used in the parsers.

File: main/parse.h (hunk 1 above)
Comment (masatake) [inline main/parse.h]: (Fixed in my private branch. )

File: main/parse.c (hunk 2 above)
Comment (masatake) [inline main/parse.c]: ... I have no option but some people may think " 0 <= language < LanguageCount  " is easier to read. See another Assert in main/parse.c. All of the take the same style.

File: main/parse.c (hunk 3 above)
Comment (masatake) [inline main/parse.c]: This is really one of my bad. (Another one  is the usage of const.)

When I started working on fishman/ctags, I carelessly introduced many name with underscore.
> ~/var/ctags-github/main% grep _ *.c *.h | grep -v define | grep -v undef | grep -v TAG| grep -v ifdef | grep -v off_t | grep -v readtags  | grep -v endif| grep -v size_t | grep -v ifndef | grep -v '[A-Z]_[A-Z]' | wc -l
650

I introduced so many areas. I mixed the naming styles too many areas.
The unified style in all source code is better but I think it is too late to fix.

**Comment by masatake** [comment]
How about error messages, long flags syntax introduced to --langdef, and documents?

File: main/parse.c (hunk 3 above)
Comment (vhda) [inline main/parse.c]: What is behind us, is behind us.
But can we agree on a naming style and follow it from now on?
At least in this specific case, that code really stands out when compared to the lines above and below!

**Comment by vhda** [comment]
I'll leave that more in depth analysis for @b4n =)

File: main/parse.c (hunk 3 above)
Comment (masatake) [inline main/parse.c]: Loog at lregex.c, I use `_` for flag related callback functions.
So the name lang_def_flag_file_kind_long is natural for me.

How about following rules?
About file static varaibles and functions we can choose any style.
About extern functions we should use camel style.

> $  grep _  *.h | grep -v define | grep -v undef | grep -v TAG| grep -v ifdef | grep -v off_t | grep -v readtags | grep -v endif| grep -v size_t | grep -v ifndef | grep -v '[A-Z]_[A-Z]' | wc -l
 grep _  *.h | grep -v define | grep -v undef | grep -v TAG| grep -v ifdef | grep -v off_t | grep -v readtags | grep -v endif| grep -v size_t | grep -v ifndef | grep -v '[A-Z]_[A-Z]' | wc -l
44

Most of all 44 hits are introduced by me. They can be fixed.

For field name, I would like to use `_`.

Do the same against *.c in parsers directlry, I got 2368 hits.

Comment (masatake) [inline main/parse.c]: O.K. I will remove _ from main/*.h.

 > $ cd main; grep _ *.h | grep -v #  | grep -v size_t | grep -v ARG_ | grep -v DEBUG_ | grep -v fpos_t | grep -v _QUOTE | grep -v RESCAN_ | grep -v EX_ | grep -v PCOPROC_ | grep -v CONST | grep -v PARSER_LIST | grep -v OPTION | grep -v SO_ | grep -v METHOD_ | grep -v __ | grep -v jmp_buf | grep -v SYMBOL

All underscores are introduced by me.

Comment (vhda) [inline main/parse.c]: Thanks! That makes the code much more consistent.

Comment (masatake) [inline main/parse.c]: Updated rules are:
* function name, type names and struct memeber name in *.h should not use underscore.
* parameter names in function decl in *.h can use underscore.
* in *.c file underscore can be used.

**Pull request merged**